File: backend/routes/time_tracking.py
from fastapi import APIRouter, HTTPException, status, Depends, Query, UploadFile, File
from typing import List, Optional
from datetime import datetime, timedelta, date
from ..models.time_tracking import TimeEntry, TimeEntryCreate, TimeEntryUpdate, TimeEntryManual, ActivityData, Screenshot
from ..models.user import User
from ..auth.dependencies import get_current_user, require_admin_or_manager
from ..database.mongodb import DatabaseOperations
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/time-tracking", tags=["time tracking"])

@router.post("/start", response_model=TimeEntry)
async def start_time_tracking(
    entry_data: TimeEntryCreate,
    current_user: User = Depends(get_current_user)
):
    """Start time tracking for a project/task"""
    try:
        # Check if user has any active time entries
        active_entry = await DatabaseOperations.get_document(
            "time_entries",
            {"user_id": current_user.id, "end_time": None}
        )
        
        if active_entry:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="You already have an active time entry. Please stop it first."
            )
        
        # Verify project exists
        project_data = await DatabaseOperations.get_document("projects", {"id": entry_data.project_id})
        if not project_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Project not found"
            )
        
        # Verify task exists if provided
        if entry_data.task_id:
            task_data = await DatabaseOperations.get_document("tasks", {"id": entry_data.task_id})
            if not task_data:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Task not found"
                )
        
        time_entry = TimeEntry(
            user_id=current_user.id,
            project_id=entry_data.project_id,
            task_id=entry_data.task_id,
            start_time=datetime.utcnow(),
            description=entry_data.description
        )
        
        await DatabaseOperations.create_document("time_entries", time_entry.dict())
        
        # Update user status to active
        await DatabaseOperations.update_document(
            "users",
            {"id": current_user.id},
            {"status": "active", "last_active": datetime.utcnow()}
        )
        
        return time_entry
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Start time tracking error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to start time tracking"
        )

@router.post("/stop/{entry_id}", response_model=TimeEntry)
async def stop_time_tracking(
    entry_id: str,
    current_user: User = Depends(get_current_user)
):
    """Stop time tracking"""
    try:
        entry_data = await DatabaseOperations.get_document(
            "time_entries",
            {"id": entry_id, "user_id": current_user.id}
        )
        
        if not entry_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Time entry not found"
            )
        
        if entry_data.get("end_time"):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Time entry already stopped"
            )
        
        end_time = datetime.utcnow()
        start_time = entry_data["start_time"]
        duration = int((end_time - start_time).total_seconds())
        
        update_data = {
            "end_time": end_time,
            "duration": duration
        }
        
        await DatabaseOperations.update_document(
            "time_entries",
            {"id": entry_id},
            update_data
        )
        
        # Update project hours
        await DatabaseOperations.update_document(
            "projects",
            {"id": entry_data["project_id"]},
            {"$inc": {"hours_tracked": duration / 3600}}
        )
        
        # Get updated entry
        updated_entry = await DatabaseOperations.get_document("time_entries", {"id": entry_id})
        return TimeEntry(**updated_entry)
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Stop time tracking error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to stop time tracking"
        )

@router.get("/active", response_model=Optional[TimeEntry])
async def get_active_time_entry(current_user: User = Depends(get_current_user)):
    """Get current user's active time entry"""
    try:
        entry_data = await DatabaseOperations.get_document(
            "time_entries",
            {"user_id": current_user.id, "end_time": None}
        )
        
        if not entry_data:
            return None
        
        return TimeEntry(**entry_data)
        
    except Exception as e:
        logger.error(f"Get active time entry error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get active time entry"
        )

@router.get("/entries", response_model=List[TimeEntry])
async def get_time_entries(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    project_id: Optional[str] = None,
    current_user: User = Depends(get_current_user)
):
    """Get time entries for current user"""
    try:
        query = {"user_id": current_user.id}
        
        if project_id:
            query["project_id"] = project_id
        
        if start_date or end_date:
            date_query = {}
            if start_date:
                date_query["$gte"] = datetime.combine(start_date, datetime.min.time())
            if end_date:
                date_query["$lte"] = datetime.combine(end_date, datetime.max.time())
            query["start_time"] = date_query
        
        entries_data = await DatabaseOperations.get_documents(
            "time_entries",
            query,
            sort=[("start_time", -1)],
            skip=skip,
            limit=limit
        )
        
        return [TimeEntry(**entry) for entry in entries_data]
        
    except Exception as e:
        logger.error(f"Get time entries error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get time entries"
        )

@router.post("/manual", response_model=TimeEntry)
async def create_manual_time_entry(
    entry_data: TimeEntryManual,
    current_user: User = Depends(get_current_user)
):
    """Create manual time entry"""
    try:
        # Verify project exists
        project_data = await DatabaseOperations.get_document("projects", {"id": entry_data.project_id})
        if not project_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Project not found"
            )
        
        # Calculate duration
        duration = int((entry_data.end_time - entry_data.start_time).total_seconds())
        
        if duration <= 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="End time must be after start time"
            )
        
        time_entry = TimeEntry(
            user_id=current_user.id,
            project_id=entry_data.project_id,
            task_id=entry_data.task_id,
            start_time=entry_data.start_time,
            end_time=entry_data.end_time,
            duration=duration,
            description=entry_data.description,
            is_manual=True
        )
        
        await DatabaseOperations.create_document("time_entries", time_entry.dict())
        
        # Update project hours
        await DatabaseOperations.update_document(
            "projects",
            {"id": entry_data.project_id},
            {"$inc": {"hours_tracked": duration / 3600}}
        )
        
        return time_entry
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Create manual time entry error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create manual time entry"
        )

@router.put("/entries/{entry_id}", response_model=TimeEntry)
async def update_time_entry(
    entry_id: str,
    entry_update: TimeEntryUpdate,
    current_user: User = Depends(get_current_user)
):
    """Update time entry"""
    try:
        entry_data = await DatabaseOperations.get_document(
            "time_entries",
            {"id": entry_id, "user_id": current_user.id}
        )
        
        if not entry_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Time entry not found"
            )
        
        update_data = entry_update.dict(exclude_unset=True)
        
        if update_data:
            await DatabaseOperations.update_document(
                "time_entries",
                {"id": entry_id},
                update_data
            )
        
        # Get updated entry
        updated_entry = await DatabaseOperations.get_document("time_entries", {"id": entry_id})
        return TimeEntry(**updated_entry)
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update time entry error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update time entry"
        )

@router.post("/activity", response_model=ActivityData)
async def record_activity(
    activity: ActivityData,
    current_user: User = Depends(get_current_user)
):
    """Record activity data"""
    try:
        activity.user_id = current_user.id
        
        await DatabaseOperations.create_document("activity_data", activity.dict())
        
        return activity
        
    except Exception as e:
        logger.error(f"Record activity error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to record activity"
        )

@router.post("/screenshot", response_model=Screenshot)
async def upload_screenshot(
    time_entry_id: str,
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user)
):
    """Upload screenshot"""
    try:
        # Verify time entry belongs to user
        entry_data = await DatabaseOperations.get_document(
            "time_entries",
            {"id": time_entry_id, "user_id": current_user.id}
        )
        
        if not entry_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Time entry not found"
            )
        
        # In a real implementation, you would save the file to cloud storage
        # For now, we'll just create a mock URL
        screenshot_url = f"/screenshots/{current_user.id}/{time_entry_id}/{file.filename}"
        
        screenshot = Screenshot(
            user_id=current_user.id,
            time_entry_id=time_entry_id,
            url=screenshot_url
        )
        
        await DatabaseOperations.create_document("screenshots", screenshot.dict())
        
        # Add screenshot to time entry
        await DatabaseOperations.update_document(
            "time_entries",
            {"id": time_entry_id},
            {"$push": {"screenshots": screenshot_url}}
        )
        
        return screenshot
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Upload screenshot error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to upload screenshot"
        )

@router.get("/reports/daily")
async def get_daily_report(
    date: Optional[date] = None,
    current_user: User = Depends(get_current_user)
):
    """Get daily time tracking report"""
    try:
        if not date:
            date = datetime.utcnow().date()
        
        start_datetime = datetime.combine(date, datetime.min.time())
        end_datetime = datetime.combine(date, datetime.max.time())
        
        # Get time entries for the day
        entries_data = await DatabaseOperations.get_documents(
            "time_entries",
            {
                "user_id": current_user.id,
                "start_time": {"$gte": start_datetime, "$lte": end_datetime}
            }
        )
        
        total_hours = sum(entry.get("duration", 0) for entry in entries_data) / 3600
        
        # Group by project
        projects = {}
        for entry in entries_data:
            project_id = entry["project_id"]
            if project_id not in projects:
                project_data = await DatabaseOperations.get_document("projects", {"id": project_id})
                projects[project_id] = {
                    "project_name": project_data["name"] if project_data else "Unknown",
                    "hours": 0,
                    "entries": 0
                }
            projects[project_id]["hours"] += entry.get("duration", 0) / 3600
            projects[project_id]["entries"] += 1
        
        return {
            "date": date,
            "total_hours": total_hours,
            "projects": projects,
            "entries_count": len(entries_data)
        }
        
    except Exception as e:
        logger.error(f"Get daily report error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get daily report"
        )

@router.get("/reports/team")
async def get_team_time_report(
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    current_user: User = Depends(require_admin_or_manager)
):
    """Get team time tracking report"""
    try:
        if not start_date:
            start_date = (datetime.utcnow() - timedelta(days=7)).date()
        if not end_date:
            end_date = datetime.utcnow().date()
        
        start_datetime = datetime.combine(start_date, datetime.min.time())
        end_datetime = datetime.combine(end_date, datetime.max.time())
        
        # Aggregate team data
        pipeline = [
            {
                "$match": {
                    "start_time": {"$gte": start_datetime, "$lte": end_datetime}
                }
            },
            {
                "$group": {
                    "_id": "$user_id",
                    "total_hours": {"$sum": "$duration"},
                    "entries_count": {"$sum": 1},
                    "projects": {"$addToSet": "$project_id"}
                }
            }
        ]
        
        team_data = await DatabaseOperations.aggregate("time_entries", pipeline)
        
        # Get user details
        for user_data in team_data:
            user_info = await DatabaseOperations.get_document("users", {"id": user_data["_id"]})
            user_data["user_name"] = user_info["name"] if user_info else "Unknown"
            user_data["total_hours"] = user_data["total_hours"] / 3600
        
        return {
            "start_date": start_date,
            "end_date": end_date,
            "team_data": team_data
        }
        
    except Exception as e:
        logger.error(f"Get team time report error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get team time report"
        )

File: backend/routes/time-tracking.js
// backend/routes/time-tracking.js - FIXED with proper error handling and validation
const express = require("express");
const { v4: uuidv4 } = require("uuid");
const TimeEntry = require("../models/TimeEntry");
const Task = require("../models/Task");
const Project = require("../models/Project");
const User = require("../models/User");
const { authMiddleware } = require("../middleware/auth");

const router = express.Router();

// Apply auth middleware to all routes
router.use(authMiddleware);

// DIAGNOSTIC ROUTE - Add this after router.use(authMiddleware);
router.post("/diagnose", async (req, res) => {
  const diagnostics = {
    timestamp: new Date(),
    checks: [],
    errors: [],
    success: false,
    user: req.user
      ? {
          id: req.user.id,
          organizationId: req.user.organizationId,
          role: req.user.role,
        }
      : null,
    requestBody: req.body,
  };

  try {
    // Check 1: Authentication
    if (!req.user) {
      diagnostics.errors.push("No user in request");
      diagnostics.checks.push({
        name: "Authentication",
        status: "❌ FAIL",
        details: "req.user is undefined",
      });
    } else {
      diagnostics.checks.push({
        name: "Authentication",
        status: "✅ PASS",
        details: `User ID: ${req.user.id}, Role: ${req.user.role}`,
      });
    }

    // Check 2: Organization
    if (!req.user?.organizationId) {
      diagnostics.errors.push("No organization ID");
      diagnostics.checks.push({
        name: "Organization",
        status: "❌ FAIL",
        details: "req.user.organizationId is undefined",
      });
    } else {
      diagnostics.checks.push({
        name: "Organization",
        status: "✅ PASS",
        details: `Organization ID: ${req.user.organizationId}`,
      });
    }

    // Check 3: Models
    try {
      const TestTimeEntry = require("../models/TimeEntry");
      const TestProject = require("../models/Project");
      diagnostics.checks.push({
        name: "Models Import",
        status: "✅ PASS",
        details: "TimeEntry and Project models imported successfully",
      });
    } catch (error) {
      diagnostics.errors.push(`Model import error: ${error.message}`);
      diagnostics.checks.push({
        name: "Models Import",
        status: "❌ FAIL",
        details: error.message,
      });
    }

    // Check 4: Database Connection
    if (diagnostics.errors.length === 0 && req.user?.organizationId) {
      try {
        const projectCount = await Project.countDocuments({
          organizationId: req.user.organizationId,
        });
        diagnostics.checks.push({
          name: "Database Connection",
          status: "✅ PASS",
          details: `Found ${projectCount} projects in organization`,
        });
      } catch (error) {
        diagnostics.errors.push(`Database error: ${error.message}`);
        diagnostics.checks.push({
          name: "Database Connection",
          status: "❌ FAIL",
          details: error.message,
        });
      }
    }

    diagnostics.success = diagnostics.errors.length === 0;
    res.json(diagnostics);
  } catch (error) {
    res.status(500).json({
      error: "Diagnostic failed",
      details: error.message,
    });
  }
});

// POST /api/time-tracking/start - Start time tracking
router.post("/start", async (req, res) => {
  try {
    const { project_id, task_id, description } = req.body;

    console.log("🚀 Starting time tracking:", {
      user: req.user.id,
      project_id,
      task_id,
      description,
      organizationId: req.user.organizationId,
    });

    // Get default project if none provided
    let actualProjectId = project_id;
    if (!actualProjectId) {
      console.log("🔍 No project provided, finding default project...");
      const defaultProject = await Project.findOne({
        organizationId: req.user.organizationId,
        status: "active"
      }).sort({ createdAt: 1 }); // Get oldest active project as default
      
      if (defaultProject) {
        actualProjectId = defaultProject.id;
        console.log(`🎯 Using default project: ${defaultProject.name} (${actualProjectId})`);
      } else {
        return res.status(400).json({
          error: "No active projects found. Please create a project first.",
        });
      }
    }

    // Verify project exists and user has access
    const project = await Project.findOne({
      id: actualProjectId,
      organizationId: req.user.organizationId,
    });

    if (!project) {
      return res.status(404).json({
        error: "Project not found or you don't have access to it",
      });
    }

    // Check if user has access to this project
    const hasProjectAccess =
      req.user.role === "admin" ||
      req.user.role === "manager" ||
      project.members.includes(req.user.id) ||
      project.manager === req.user.id;

    if (!hasProjectAccess) {
      return res.status(403).json({
        error: "You don't have access to track time on this project",
      });
    }

    // Verify task if provided
    let task = null;
    if (task_id) {
      task = await Task.findOne({
        id: task_id,
        organizationId: req.user.organizationId,
        project_id: actualProjectId,
      });

      if (!task) {
        return res.status(404).json({
          error: "Task not found or doesn't belong to the selected project",
        });
      }

      // Check if user can track time on this task
      const hasTaskAccess =
        req.user.role === "admin" ||
        req.user.role === "manager" ||
        task.assignee_id === req.user.id ||
        task.createdBy === req.user.id ||
        task.watchers.includes(req.user.id);

      if (!hasTaskAccess) {
        return res.status(403).json({
          error: "You don't have permission to track time on this task",
        });
      }
    }

    // Check if user already has an active entry
    const activeEntry = await TimeEntry.findOne({
      organizationId: req.user.organizationId,
      user_id: req.user.id,
      end_time: null,
    });

    if (activeEntry) {
      return res.status(400).json({
        error: "You already have an active time entry. Please stop it first.",
        activeEntry: {
          id: activeEntry.id,
          project_id: activeEntry.project_id,
          start_time: activeEntry.start_time,
          description: activeEntry.description,
        },
      });
    }

    // Create new time entry
    const timeEntry = new TimeEntry({
      id: uuidv4(),
      organizationId: req.user.organizationId,
      user_id: req.user.id,
      project_id: actualProjectId,
      task_id: task_id || null,
      description: description?.trim() || "General work time",
      start_time: new Date(),
      billable: true, // Default to billable
      hourly_rate: req.user.billing?.hourlyRate || 0,
    });

    await timeEntry.save();

    // Update task status to in_progress if task is provided and currently todo
    if (task && task.status === "todo") {
      task.status = "in_progress";
      task.startedAt = new Date();
      await task.save();
      console.log(`📝 Task status updated to in_progress: ${task.title}`);
    }

    // Update task time tracking info
    if (task) {
      task.timeTracking.isActive = true;
      task.timeTracking.activeEntryId = timeEntry.id;
      task.timeTracking.lastStarted = new Date();
      await task.save();
    }

    console.log("✅ Time tracking started successfully:", {
      entryId: timeEntry.id,
      project: project.name,
      task: task?.title,
      user: req.user.name,
    });

    res.status(201).json({
      message: "Time tracking started successfully",
      id: timeEntry.id,
      project_id: timeEntry.project_id,
      task_id: timeEntry.task_id,
      description: timeEntry.description,
      start_time: timeEntry.start_time,
      project_name: project.name,
      task_title: task?.title,
      auto_selected: !project_id, // Indicate if project was auto-selected
    });
  } catch (error) {
    console.error("❌ Start tracking error:", error);

    if (error.name === "ValidationError") {
      const errors = Object.values(error.errors).map((err) => err.message);
      return res.status(400).json({
        error: "Validation failed",
        errors,
      });
    }

    res.status(500).json({
      error: "Failed to start time tracking",
      details:
        process.env.NODE_ENV === "development" ? error.message : undefined,
    });
  }
});

// POST /api/time-tracking/stop/:entryId - Stop time tracking
router.post("/stop/:entryId", async (req, res) => {
  try {
    const { entryId } = req.params;

    console.log("🛑 Stopping time tracking:", {
      entryId,
      user: req.user.id,
      organizationId: req.user.organizationId,
    });

    // Find the active time entry
    const timeEntry = await TimeEntry.findOne({
      id: entryId,
      organizationId: req.user.organizationId,
      user_id: req.user.id,
      end_time: null,
    });

    if (!timeEntry) {
      return res.status(404).json({
        error: "Active time entry not found",
      });
    }

    // Calculate duration and stop the entry
    const endTime = new Date();
    const duration = Math.round((endTime - timeEntry.start_time) / 1000); // in seconds
    const totalAmount = (duration / 3600) * (timeEntry.hourly_rate || 0);

    timeEntry.end_time = endTime;
    timeEntry.duration = duration;
    timeEntry.total_amount = totalAmount;

    await timeEntry.save();

    // Update task time tracking info if applicable
    if (timeEntry.task_id) {
      const task = await Task.findOne({
        id: timeEntry.task_id,
        organizationId: req.user.organizationId,
      });

      if (task) {
        task.timeTracking.isActive = false;
        task.timeTracking.activeEntryId = null;
        task.timeTracking.totalTracked += duration;

        // Update actual hours if task has updateTimeTracking method
        if (task.updateTimeTracking) {
          await task.updateTimeTracking();
        }

        await task.save();

        console.log(
          `📝 Task time tracking updated: ${task.title} (+${Math.round(
            duration / 60
          )} minutes)`
        );
      }
    }

    // Update project stats if project has updateStats method
    try {
      const project = await Project.findOne({
        id: timeEntry.project_id,
        organizationId: req.user.organizationId,
      });
      if (project && project.updateStats) {
        await project.updateStats();
      }
    } catch (projectError) {
      console.warn("Failed to update project stats:", projectError);
    }

    console.log("✅ Time tracking stopped successfully:", {
      entryId: timeEntry.id,
      duration: `${Math.round(duration / 60)} minutes`,
      totalAmount: `$${totalAmount.toFixed(2)}`,
    });

    res.json({
      message: "Time tracking stopped successfully",
      id: timeEntry.id,
      duration: duration,
      total_amount: totalAmount,
      start_time: timeEntry.start_time,
      end_time: timeEntry.end_time,
    });
  } catch (error) {
    console.error("❌ Stop tracking error:", error);
    res.status(500).json({
      error: "Failed to stop time tracking",
      details:
        process.env.NODE_ENV === "development" ? error.message : undefined,
    });
  }
});

// GET /api/time-tracking/active - Get active time entry
router.get("/active", async (req, res) => {
  try {
    const activeEntry = await TimeEntry.findOne({
      organizationId: req.user.organizationId,
      user_id: req.user.id,
      end_time: null,
    }).sort({ start_time: -1 });

    if (!activeEntry) {
      return res.json(null);
    }

    // Get project and task info
    const [project, task] = await Promise.all([
      Project.findOne({
        id: activeEntry.project_id,
        organizationId: req.user.organizationId,
      }),
      activeEntry.task_id
        ? Task.findOne({
            id: activeEntry.task_id,
            organizationId: req.user.organizationId,
          })
        : null,
    ]);

    res.json({
      id: activeEntry.id,
      project_id: activeEntry.project_id,
      task_id: activeEntry.task_id,
      description: activeEntry.description,
      start_time: activeEntry.start_time,
      project_name: project?.name,
      task_title: task?.title,
    });
  } catch (error) {
    console.error("❌ Get active entry error:", error);
    res.status(500).json({
      error: "Failed to fetch active time entry",
    });
  }
});

// GET /api/time-tracking/entries - Get time entries
router.get("/entries", async (req, res) => {
  try {
    const {
      limit = 20,
      page = 1,
      project_id,
      task_id,
      start_date,
      end_date,
    } = req.query;

    const skip = (page - 1) * limit;
    const query = {
      organizationId: req.user.organizationId,
      user_id: req.user.id,
    };

    // Apply filters
    if (project_id) query.project_id = project_id;
    if (task_id) query.task_id = task_id;
    if (start_date) {
      query.start_time = { $gte: new Date(start_date) };
    }
    if (end_date) {
      query.start_time = {
        ...query.start_time,
        $lte: new Date(end_date),
      };
    }

    const [entries, total] = await Promise.all([
      TimeEntry.find(query)
        .sort({ start_time: -1 })
        .limit(parseInt(limit))
        .skip(skip),
      TimeEntry.countDocuments(query),
    ]);

    // Get project and task details for each entry
    const enrichedEntries = await Promise.all(
      entries.map(async (entry) => {
        const [project, task] = await Promise.all([
          Project.findOne({
            id: entry.project_id,
            organizationId: req.user.organizationId,
          }),
          entry.task_id
            ? Task.findOne({
                id: entry.task_id,
                organizationId: req.user.organizationId,
              })
            : null,
        ]);

        return {
          ...entry.toJSON(),
          project_name: project?.name,
          project_client: project?.client,
          task_title: task?.title,
        };
      })
    );

    res.json({
      entries: enrichedEntries,
      pagination: {
        total,
        page: parseInt(page),
        limit: parseInt(limit),
        pages: Math.ceil(total / limit),
      },
    });
  } catch (error) {
    console.error("❌ Get entries error:", error);
    res.status(500).json({
      error: "Failed to fetch time entries",
    });
  }
});

// POST /api/time-tracking/manual - Create manual time entry
router.post("/manual", async (req, res) => {
  try {
    const {
      project_id,
      task_id,
      description,
      start_time,
      end_time,
      billable = true,
    } = req.body;

    // Validation
    if (!project_id || !start_time || !end_time) {
      return res.status(400).json({
        error: "Project ID, start time, and end time are required",
      });
    }

    const startDate = new Date(start_time);
    const endDate = new Date(end_time);

    if (endDate <= startDate) {
      return res.status(400).json({
        error: "End time must be after start time",
      });
    }

    // Verify project access
    const project = await Project.findOne({
      id: project_id,
      organizationId: req.user.organizationId,
    });

    if (!project) {
      return res.status(404).json({
        error: "Project not found",
      });
    }

    // Verify task if provided
    if (task_id) {
      const task = await Task.findOne({
        id: task_id,
        organizationId: req.user.organizationId,
        project_id: project_id,
      });

      if (!task) {
        return res.status(404).json({
          error: "Task not found",
        });
      }
    }

    // Calculate duration
    const duration = Math.round((endDate - startDate) / 1000);
    const totalAmount = (duration / 3600) * (req.user.billing?.hourlyRate || 0);

    // Create manual time entry
    const timeEntry = new TimeEntry({
      id: uuidv4(),
      organizationId: req.user.organizationId,
      user_id: req.user.id,
      project_id,
      task_id: task_id || null,
      description: description?.trim() || "Manual time entry",
      start_time: startDate,
      end_time: endDate,
      duration,
      total_amount: totalAmount,
      billable,
      is_manual: true,
      hourly_rate: req.user.billing?.hourlyRate || 0,
    });

    await timeEntry.save();

    res.status(201).json({
      message: "Manual time entry created successfully",
      entry: timeEntry,
    });
  } catch (error) {
    console.error("❌ Create manual entry error:", error);
    res.status(500).json({
      error: "Failed to create manual time entry",
    });
  }
});

// PUT /api/time-tracking/entries/:entryId - Update time entry
router.put("/entries/:entryId", async (req, res) => {
  try {
    const { entryId } = req.params;
    const { description, billable, start_time, end_time } = req.body;

    const timeEntry = await TimeEntry.findOne({
      id: entryId,
      organizationId: req.user.organizationId,
      user_id: req.user.id,
    });

    if (!timeEntry) {
      return res.status(404).json({
        error: "Time entry not found",
      });
    }

    // Update allowed fields
    if (description !== undefined) timeEntry.description = description;
    if (billable !== undefined) timeEntry.billable = billable;

    // Handle time changes for completed entries
    if (timeEntry.end_time && start_time && end_time) {
      const newStart = new Date(start_time);
      const newEnd = new Date(end_time);

      if (newEnd <= newStart) {
        return res.status(400).json({
          error: "End time must be after start time",
        });
      }

      timeEntry.start_time = newStart;
      timeEntry.end_time = newEnd;
      timeEntry.duration = Math.round((newEnd - newStart) / 1000);
      timeEntry.total_amount =
        (timeEntry.duration / 3600) * (timeEntry.hourly_rate || 0);
    }

    await timeEntry.save();

    res.json({
      message: "Time entry updated successfully",
      entry: timeEntry,
    });
  } catch (error) {
    console.error("❌ Update entry error:", error);
    res.status(500).json({
      error: "Failed to update time entry",
    });
  }
});

// DELETE /api/time-tracking/entries/:entryId - Delete time entry
router.delete("/entries/:entryId", async (req, res) => {
  try {
    const { entryId } = req.params;

    const timeEntry = await TimeEntry.findOne({
      id: entryId,
      organizationId: req.user.organizationId,
      user_id: req.user.id,
    });

    if (!timeEntry) {
      return res.status(404).json({
        error: "Time entry not found",
      });
    }

    // Cannot delete active entries
    if (!timeEntry.end_time) {
      return res.status(400).json({
        error: "Cannot delete active time entry. Stop it first.",
      });
    }

    await TimeEntry.deleteOne({ id: entryId });

    res.json({
      message: "Time entry deleted successfully",
    });
  } catch (error) {
    console.error("❌ Delete entry error:", error);
    res.status(500).json({
      error: "Failed to delete time entry",
    });
  }
});

// GET /api/time-tracking/stats - Get time tracking statistics
router.get("/stats", async (req, res) => {
  try {
    const { period = "week" } = req.query;
    const now = new Date();
    let startDate;

    switch (period) {
      case "today":
        startDate = new Date(now.getFullYear(), now.getMonth(), now.getDate());
        break;
      case "week":
        startDate = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
        break;
      case "month":
        startDate = new Date(now.getFullYear(), now.getMonth(), 1);
        break;
      default:
        startDate = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
    }

    const stats = await TimeEntry.aggregate([
      {
        $match: {
          organizationId: req.user.organizationId,
          user_id: req.user.id,
          start_time: { $gte: startDate },
          end_time: { $ne: null },
        },
      },
      {
        $group: {
          _id: null,
          totalDuration: { $sum: "$duration" },
          totalAmount: { $sum: "$total_amount" },
          entryCount: { $sum: 1 },
          avgDuration: { $avg: "$duration" },
        },
      },
    ]);

    const result = stats[0] || {
      totalDuration: 0,
      totalAmount: 0,
      entryCount: 0,
      avgDuration: 0,
    };

    res.json({
      period,
      stats: {
        totalHours: Math.round((result.totalDuration / 3600) * 100) / 100,
        totalAmount: Math.round(result.totalAmount * 100) / 100,
        entryCount: result.entryCount,
        avgHours: Math.round((result.avgDuration / 3600) * 100) / 100,
      },
    });
  } catch (error) {
    console.error("❌ Get stats error:", error);
    res.status(500).json({
      error: "Failed to fetch time tracking statistics",
    });
  }
});

module.exports = router;